src/org/pentaho/reporting/platform/plugin/output/CachingPageableHTMLOutput.java

```diff
@@ -161,31 +178,87 @@ private IReportContent produceCacheablePages( final MasterReport report, final i
     final PageableHtmlOutputProcessor outputProcessor = (PageableHtmlOutputProcessor) proc.getOutputProcessor();
     outputProcessor.setFlowSelector( new DisplayAllFlowSelector() );
 
-    final Map<Integer, byte[]> pages = new HashMap<>();
+    final ExtendedConfiguration config = ClassicEngineBoot.getInstance().getExtendedConfig();
+    final boolean firstPageMode =
+      config.getBoolProperty( "org.pentaho.reporting.platform.plugin.output.FirstPageMode" );
+
+    //Async listener
+    final ReportProgressListener listener = ReportListenerThreadHolder.getListener();
+    if ( listener != null ) {
+      proc.addReportProgressListener( listener );
+      //Enable First Page Content available feature
+      if ( firstPageMode ) {
+        ReportListenerThreadHolder.enableFirstPageMode();
+      }
+    }
+    ReportProgressListener cacheListener = null;
     try {
       final ZipRepository targetRepository = reinitOutputTargetForCaching();
-      proc.processReport();
-      final int pageCount = proc.getLogicalPageCount();
-      final ContentLocation root = targetRepository.getRoot();
-
-      for ( final ContentEntity contentEntities : root.listContents() ) {
-        if ( contentEntities instanceof ContentItem ) {
-          final ContentItem ci = (ContentItem) contentEntities;
-          final String name = ci.getName();
-          final int pageNumber = extractPageFromName( name );
-          if ( pageNumber >= 0 ) {
-            pages.put( pageNumber, read( ci.getInputStream() ) );
+      if ( firstPageMode ) {
+        //Create cache listener to write first page when needed
+        cacheListener = new ReportProgressListener() {
+
+          @Override public void reportProcessingStarted( final ReportProgressEvent reportProgressEvent ) {
+            //ignore
           }
-        }
+
+          @Override public void reportProcessingUpdate( final ReportProgressEvent reportProgressEvent ) {
+            if ( reportProgressEvent.getActivity() == ReportProgressEvent.GENERATING_CONTENT
+              && reportProgressEvent.getPage() == 1 ) {
```

Comment:
This feature is not "firstPage" mode as in return the page with number 1. The user requests a specific page via the "accept-page" parameter, and *that* is what the user wants. 

So when you hit the page specified by "acceptedPage" *then* you can return the content of that page. Always returning the first page will break the user's expectations.